css/src/lib.rs

```diff
@@ -317,75 +250,69 @@ impl TryFrom<&[u8]> for Signature {
             return Err(Error::Length);
         }
 
-        let (bytes, header) = take(HEADER_LEN)(src.as_bytes());
-        let (bytes, vendor) = take(VENDOR_LEN)(bytes);
-        let (bytes, date) = take(DATE_LEN)(bytes);
-        let (bytes, header2) = take(HEADER2_LEN)(bytes);
-        let (bytes, swdefined) = take(SWDEFINED_LEN)(bytes);
-        let (bytes, _reserved1) = take(RESERVED1_LEN)(bytes);
-        let (bytes, modulus) = take(PUBKEY_LEN)(bytes);
-        let (bytes, exponent) = take(EXPONENT_LEN)(bytes);
-        let (bytes, signature) = take(SIGNATURE_LEN)(bytes);
-        let (bytes, miscselect) = take(MISCSELECT_LEN)(bytes);
-        let (bytes, miscmask) = take(MISCMASK_LEN)(bytes);
-        let (bytes, _reserved2) = take(RESERVED2_LEN)(bytes);
-        let (bytes, attributes) = take(ATTRIBUTES_LEN)(bytes);
-        let (bytes, attributemask) = take(ATTRIBUTEMASK_LEN)(bytes);
-        let (bytes, enclavehash) = take(MRENCLAVE_LEN)(bytes);
-        let (bytes, _reserved3) = take(RESERVED3_LEN)(bytes);
-        let (bytes, isvprodid) = take(ISVPRODID_LEN)(bytes);
-        let (bytes, isvsvn) = take(ISVSVN_LEN)(bytes);
-        let (bytes, _reserved4) = take(RESERVED4_LEN)(bytes);
-        let (bytes, q1) = take(Q1_LEN)(bytes);
-        let (_bytes, q2) = take(Q2_LEN)(bytes);
+        let (bytes, header) = take::<HEADER_LEN>()(src.as_bytes());
+        let (bytes, vendor) = take::<VENDOR_LEN>()(bytes);
+        let (bytes, date) = take::<DATE_LEN>()(bytes);
+        let (bytes, header2) = take::<HEADER2_LEN>()(bytes);
+        let (bytes, swdefined) = take::<SWDEFINED_LEN>()(bytes);
+        let (bytes, _reserved1) = take::<RESERVED1_LEN>()(bytes);
+        let (bytes, modulus) = take::<PUBKEY_LEN>()(bytes);
+        let (bytes, exponent) = take::<EXPONENT_LEN>()(bytes);
+        let (bytes, signature) = take::<SIGNATURE_LEN>()(bytes);
+        let (bytes, miscselect) = take::<MISCSELECT_LEN>()(bytes);
+        let (bytes, miscmask) = take::<MISCMASK_LEN>()(bytes);
+        let (bytes, _reserved2) = take::<RESERVED2_LEN>()(bytes);
+        let (bytes, attributes) = take::<ATTRIBUTES_LEN>()(bytes);
+        let (bytes, attributemask) = take::<ATTRIBUTEMASK_LEN>()(bytes);
+        let (bytes, enclavehash) = take::<MRENCLAVE_LEN>()(bytes);
+        let (bytes, _reserved3) = take::<RESERVED3_LEN>()(bytes);
+        let (bytes, isvprodid) = take::<ISVPRODID_LEN>()(bytes);
+        let (bytes, isvsvn) = take::<ISVSVN_LEN>()(bytes);
+        let (bytes, _reserved4) = take::<RESERVED4_LEN>()(bytes);
+        let (bytes, q1) = take::<Q1_LEN>()(bytes);
+        let (_bytes, q2) = take::<Q2_LEN>()(bytes);
+
+        let vendor = u32::from_le_bytes(vendor).try_into()?;
```

Comment:
I think all the checks, not just the `vendor` should be before the initialization of the struct that way we never have an invalid struct, however the struct itself allows for invalid field states so..